### PR TITLE
Adaptive Localization notebook

### PR DESCRIPTION
**Pull request by tommyod**
I created a notebook to understand better how Adaptive Localization works. The notebook contains the simplest problem I could think of where I expect adaptive localization to outperform standard ESMDA -- a Gauss-Linear problem with sparse structure in the forward model. I hope this notebook and problem can be used for quality assurance in the future if we e.g. experiment with thresholding, add the Lasso idea, etc.

A Gauss-linear problem with a very local structure in the forward model. Each output is determined by at most 3 parameters. The forward map has this structure:

image

## Results (correlation coefficients)

Some results with the default threshold $3 / \sqrt{N}$

### 50 parameters, 50 observations

image

### 100 parameters, 50 observations

image

### 500 parameters, 50 observations

image

## Comments on results

image

- First of all, I am happy to see that in general, the adaptive localization seems to work. It works especially well with many ensemble members.
- When `num_ensemble` <= 9, the AdaptiveESMDA posterior equals the prior. No information is used. This happens irrespective of the problem size. If you have a lot of parameters this might make sense, but consider a small-parameter case (1 parameter and 1 response being the extreme example) - it does not make sense much sense to me to not update a prior if you have 9 observations and you're trying to estimate one parameter.
- On this problem, it seems like `num_ensemble` has to surpass ~15 for adaptive localization to correlate better with true parameter values compared to ESMDA.

### CODE REVIEW

File: docs/source/Adaptive.py [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+# Initial ensemble X ~ N(0, prior_std) and diagonal covariance with ones
+X = rng.normal(size=(num_parameters, num_ensemble)) * prior_std
+
+# Covariance matches the noise added to observations above
+covariance = np.ones(num_observations)  # N(0, 1) covariance
+
+# %% [markdown]
+# ## Solve the maximum likelihood problem
+#
+# We can solve $Ax = b$, where $b$ is the observations,
```
Comment (dafeda) [inline docs/source/Adaptive.py:109]:
```suggestion
# We can solve $Ax = b$, where $b$ are the observations,
```

File: docs/source/Adaptive.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+plt.xlabel("Parameter index")
+plt.ylabel("Parameter value")
+plt.grid(True, ls="--", zorder=0, alpha=0.33)
+plt.legend()
+plt.show()
+
+# %% [markdown]
+# ## Solve using ESMDA
+#
+# We crease an `ESMDA` instance and solve the Guass-linear problem.
```
Comment (dafeda) [inline docs/source/Adaptive.py:130]:
```suggestion
# We create an `ESMDA` instance and solve the Guass-linear problem.
```

File: docs/source/Adaptive.py [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+plt.xlabel("Parameter index")
+plt.ylabel("Parameter value")
+plt.grid(True, ls="--", zorder=0, alpha=0.33)
+plt.legend()
+plt.show()
+
+# %% [markdown]
+# ## Solve using AdaptiveESMDA
+#
+# We crease an `AdaptiveESMDA` instance and solve the Guass-linear problem.
```
Comment (dafeda) [inline docs/source/Adaptive.py:173]:
```suggestion
# We create an `AdaptiveESMDA` instance and solve the Guass-linear problem.
```

**Comment by tommyod** [comment]
Added the prior too:
image

**Review by Blunde1 — APPROVED** [review]
Very nice! Comment on natural baseline on slack. Approved, LGTM.

**Pull request merged**